fvm/env.go

```diff
@@ -11,3 +26,120 @@ type Environment interface {
 	VM() *VirtualMachine
 	runtime.Interface
 }
+
+// Parts of the environment that are common to all transaction and script
+// executions.
+type commonEnv struct {
+	ctx           Context
+	sth           *state.StateHolder
+	vm            *VirtualMachine
+	programs      *handler.ProgramsHandler
+	accounts      state.Accounts
+	accountKeys   *handler.AccountKeyHandler
+	contracts     *handler.ContractHandler
+	uuidGenerator *state.UUIDGenerator
+	metrics       *handler.MetricsHandler
+	logs          []string
+	rng           *rand.Rand
+	traceSpan     opentracing.Span
+}
+
+func (env *commonEnv) Context() *Context {
+	return &env.ctx
+}
+
+func (env *commonEnv) VM() *VirtualMachine {
+	return env.vm
+}
+
+func (env *commonEnv) seedRNG(header *flow.Header) {
+	// Seed the random number generator with entropy created from the block
+	// header ID. The random number generator will be used by the UnsafeRandom
+	// function.
+	id := header.ID()
+	source := rand.NewSource(int64(binary.BigEndian.Uint64(id[:])))
+	env.rng = rand.New(source)
+}
+
+func (env *commonEnv) isTraceable() bool {
+	return env.ctx.Tracer != nil && env.traceSpan != nil
+}
+
+func (env *commonEnv) ImplementationDebugLog(message string) error {
+	env.ctx.Logger.Debug().Msgf("Cadence: %s", message)
+	return nil
+}
+
+func (env *commonEnv) RecordTrace(operation string, location common.Location, duration time.Duration, logs []opentracing.LogRecord) {
+	if !env.isTraceable() {
+		return
+	}
+	if location != nil {
+		if logs == nil {
+			logs = make([]opentracing.LogRecord, 0, 1)
```

Comment:
The only line of code with actual change.  the original was `logs = make([]opentracing.LogRecord, 0)`